include/pybind11/pybind11.h

```diff
@@ -50,13 +50,25 @@ class cpp_function : public function {
         initialize(f, f, extra...);
     }
 
+#if !defined(_MSC_VER)
     /// Construct a cpp_function from a lambda function (possibly with internal state)
     template <typename Func, typename... Extra,
              typename FuncType = typename detail::remove_class<decltype(&std::remove_reference<Func>::type::operator())>::type>
     cpp_function(Func &&f, const Extra&... extra) {
```

Comment:
If you'd like to avoid the preprocessor, an alternative solution would be to do SFINAE without `decltype` and move `FuncType` inside the constructor. I've tried the following purely `<type_traits>`-base SFINAE condition and it works well with VS2017:
```
detail::enable_if_t<!std::is_pointer<F>::value && !std::is_member_pointer<F>::value>>
```